bpf/lib/nodeport.h

```diff
@@ -710,51 +698,91 @@ static __always_inline bool nodeport_uses_dsr4(const struct ipv4_ct_tuple *tuple
 }
 
 static __always_inline bool nodeport_nat_ipv4_needed(struct __ctx_buff *ctx,
-						     __be32 addr, int dir)
+						     __be32 addr, int dir,
+						     bool *from_endpoint __maybe_unused,
+						     const bool encap __maybe_unused)
 {
 	void *data, *data_end;
 	struct iphdr *ip4;
 
 	if (!revalidate_data(ctx, &data, &data_end, &ip4))
 		return false;
-#ifdef ENABLE_DSR_HYBRID
-	if (nodeport_uses_dsr(ip4->protocol))
-		return false;
-#endif /* ENABLE_DSR_HYBRID */
 	/* Basic minimum is to only NAT when there is a potential of
 	 * overlapping tuples, e.g. applications in hostns reusing
 	 * source IPs we SNAT in node-port.
 	 */
-	if (dir == NAT_DIR_EGRESS)
-		return ip4->saddr == addr;
-	else
+	if (dir == NAT_DIR_EGRESS) {
+		if (ip4->saddr == addr) {
+			return true;
+		}
+#ifdef ENABLE_MASQUERADE
+		struct endpoint_info *ep;
+
+		/* Do not MASQ when this function is executed from bpf_overlay
+		 * (encap=true denotes this fact). Otherwise, a packet will be
+		 * SNAT'd to cilium_host IP addr. */
+		if (encap)
+			return false;
+
+		/* TODO(brb) Optimization: use skb mark to determine if src is
+		 * local endpoint */
+		if ((ep = __lookup_ip4_endpoint(ip4->saddr)) != NULL &&
+		    !(ep->flags & ENDPOINT_F_HOST)) {
+			struct remote_endpoint_info *info;
+			*from_endpoint = true;
+
+			info = ipcache_lookup4(&IPCACHE_MAP, ip4->daddr,
+					       V4_CACHE_KEY_LEN);
+			if (info != NULL) {
+#ifdef ENABLE_IP_MASQ_AGENT
+				struct lpm_v4_key pfx;
+				pfx.lpm.prefixlen = 32;
+				memcpy(pfx.lpm.data, &ip4->daddr, sizeof(pfx.addr));
+				if (map_lookup_elem(&IP_MASQ_AGENT_IPV4, &pfx))
+					return false;
+#endif
+				if (info->sec_label == WORLD_ID)
+					return true;
+#ifdef ENCAP_IFINDEX
+				/* In the tunnel mode, a packet from a local ep
+				 * to a remote node is not encap'd, and is sent
+				 * via a native dev. Therefore, such packet has
+				 * to be MASQ'd. Otherwise, it will be dropped
+				 * either by underlying network or by the remote
+				 * node */
```

Comment:
nit: I think it's usually more about the replies being dropped because there's no route back? This paragraph I think implies that the forward direction packet may be dropped, but I don't think it's common to drop packets with unroutable source.

Reply:
It can be dropped because of either:

- An underlying network (e.g. by default in AWS) drops packets which src is from an unknown subnet.
- If `rp_filter=1` is set for a native dev of the remote node, then the packet will be dropped by the remote node, because it has a route for the src IP of the packet via a different device (i.e. `cilium_vxlan`).